### ISSUE TITLE
Project 2 updating the availability of self checkout

### ISSUE THREAD

**Issue by yixin567** [nus-cs2030/2021-s1#468]
# Description 
Describe the question  
Can i ask for my human server, i update the status on whether their queue is full using a boolean but for SelfCheckOut, how can i update the status since they share the same queue?  

Any suggestion is appreciated heh!!
Thank you!! 
## Topic: 
Specific Topics this question is related to (If Any) 

## Screenshots (if any): 
Insert Images here if necessary

**Comment by Josephaedan** [comment]
You can just define a method with Boolean return type called isFull() that returns the queue.size() >= maxQueueSize. The self checkouts all share the same queue so running this method on all of them will produce the same result.

**Comment by yixin567** [comment]
@Josephaedan does that mean that for my humanServer, i would need to use the same method too cos they both inherit from the super class Server so when i use shop.find() at the arriveEvent, it is referring to both the human and the  selfCheck

**Comment by glennljs** [comment]
@yixin567 You can define the method as an abstract method in your super class (I assume it's an abstract class) then you override it in your human/selfCheck respectively to achieve the different results and it should work via polymorphism.

**Comment by yixin567** [comment]
oh my server is  a class itself, not an abstract class @glennljs

**Comment by glennljs** [comment]
@yixin567 well since you have both human and selfCheck subclasses then you should not be instantiating your Server class so you can change it into an abstract class!

**Comment by yixin567** [comment]
@glennljs but for the previous testcase such as the test case 2, they need to instantiate a Server class

**Comment by glennljs** [comment]
@yixin567 you can change your current Server class to a different name such as "Counter" and then change ur humanServer to Server and it should work.

**Comment by qiiwen** [comment]
hello! may i ask more on how u implement the selfcheck queue?

**Comment by yixin567** [comment]
@qiiwen i used static but im not sure it its correct though...

**Comment by Tetrerox** [comment]
@qiiwen  Not the best method but for my SelfCheck queue, I have a static list whose only element is the number of waiting customers in the shared queue. (Used a list so that I can mutate the number of waiting customers) You can consider having static methods that increase or decrease the shared queue length. I have an overloaded constructor in my Server class to create a 'SelfCheck' server.

**Comment by qiiwen** [comment]
hmm, i implemented it as an immutable queue just like human server. so once any of the self checkout queues change in WaitEvent or DoneEvent, i update all other queues by constructing new selfCheck servers with the new queue passed in as parameter. so u can constantly update the status of all self check servers with the same condition used in your human server once the 'shared' queue changes. hope this helps! let me know more details if needed!

**Comment by joelwongjy** [comment]
I think most people used a List. I used an integer for both Server and SelfCheckout and it works fine. The static variable does not have to be final.

**Comment by Tetrerox** [comment]
@qiiwen Thanks for the input. I managed to get mine to work for lvl6 but the only issue I have right now is how to get the events to print 'by self-check' instead of 'by server' because I can't access the shop from the event itself to find out the type of my server.

**Comment by Tetrerox** [comment]
Oh static variables don't have to be final in codecrunch? I thought it needed to be so I used a list instead lol

**Comment by joelwongjy** [comment]
> @qiiwen Thanks for the input. I managed to get mine to work for lvl6 but the only issue I have right now is how to get the events to print 'by self-check' instead of 'by server' because I can't access the shop from the event itself to find out the type of my server. 

What I did is I have an abstract class called Counter for Server and SelfCheckout, and I created a method .print() that returns a String " self-checkout " + this.id. So in each event I'm passing along the counter and calling counter.print()

Take note you should not modify the toString() method as it is still needed to pass the earliest test cases.

**Comment by joelwongjy** [comment]
> Oh static variables don't have to be final in codecrunch? I thought it needed to be so I used a list instead lol

Yup it doesn't haha not many people know that I think but it passed on codecrunch no problems.

**Comment by Tetrerox** [comment]
@joelwongjy Oh I see. Looks like my issue is that i passed in the server identifier through the events, instead of the server itself. I will give your method a go. Thanks!

**Comment by yixin567** [comment]
@joelwongjy hi! can i ask did u add to ur static queue in the self checkout server like this? because i feel that mine is not adding even though i added put this when the customer is in the waiting queue

![image](https://user-images.githubusercontent.com/66197880/99895771-ec9f3100-2cc4-11eb-9bdd-6e0487bd0954.png)

**Comment by joelwongjy** [comment]
I modified it within SelfCheckout. Inside the SelfCheckout.serve method, where you can pass in customer and do this.list.add(customer). This kinda breaks the no side effects rule but it works best. I also have a method in SelfCheckout that compares size with qmax and returns a Boolean that helps with the tell-don't-ask principle.

Your method doesn't seem like it's adding the customer because it's effectively creating a copy of the static list. The only way to modify a static variable outside the class is if your class is static.
